File: Fitness-ai-frontend/src/types/pose-scoring.ts
export type PoseScoringStatus = "scored" | "unsupported";

export interface PoseScoringPhase {
  phase: string;
  frame_index: number;
  timestamp_ms: number;
  angle: number;
}

export type PoseErrorSeverity = "none" | "minor" | "major";

export interface PoseScoringError {
  code: string;
  label: string;
  severity: PoseErrorSeverity;
  feedback: string;
  evidence: Record<string, unknown>;
}

export interface PoseScoringMetrics {
  valid_frames?: number;
  min_angle?: number;
  max_angle?: number;
  angle_range?: number;
  phases?: PoseScoringPhase[];
  errors?: PoseScoringError[];
  [key: string]: unknown;
}

export interface PoseScoringResult {
  record_id: number;
  status: PoseScoringStatus;
  applied: boolean;
  exercise_type: string | null;
  score: number | null;
  count: number | null;
  confidence: number | null;
  feedback: string[];
  metrics: PoseScoringMetrics;
}

File: app/schemas/pose_scoring.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class PoseScoringRequest(BaseModel):
    apply: bool = False


class PoseScoringResponse(BaseModel):
    record_id: int
    status: Literal["scored", "unsupported"]
    applied: bool = False
    exercise_type: Optional[str] = None
    score: Optional[float] = None
    count: Optional[int] = None
    confidence: Optional[float] = None
    feedback: List[str] = Field(default_factory=list)
    metrics: Dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(from_attributes=True)
